### DISCUSSION TITLE
Using python decorators

### DISCUSSION THREAD

**Discussion by b-simjoo** [python-telegram-bot/python-telegram-bot#4474] [Feature Suggestions]
I'm using your module for a while and I was thinking of add decorators to make writing handlers more fun. for example adding a command handler looks like this:
```python
@bot.command
def start(u: Update, c: ContextTypes.DEFAULT_TYPE):
   ...
   
@bot.command("id")
def get_user_id(u: Update, c: ContextTypes.DEFAULT_TYPE):
    ...
```

or conversation handlers could be written like this:
```python
ORDERING=object()
PAYMENT=object()

@bot.conversation_handler
def order(u: Update, c: ContextTypes.DEFAULT_TYPE):    # order command starts the conversation handler
    ...

@order.state(ORDERING)
def ordering(u: Update, c: ContextTypes.DEFAULT_TYPE):
    ...

@order.state(PAYMENT)
def order_payment(u: Update, c: ContextTypes.DEFAULT_TYPE):
    ...

@order.fallback
def some_function(u: Update, c: ContextTypes.DEFAULT_TYPE):
    ...
```

What do you think? If you agree with that I could begin working on it and do PRs

**Comment by Poolitzer** [comment]
We had this discussion previously and think that decorators hide explicit function calls in a black magic kinda box, which we don't want to have users get used too.

So I would vote against having this in the main library, but a ptbcontrib module would be fine

**Reply by b-simjoo** [reply]
I agree with that, I like that users be free to choose the style they want and have access to different approaches.

**Comment by Bibo-Joshi** [comment]
See also #899 for discussion

**Comment by b-simjoo** [comment]
https://github.com/python-telegram-bot/python-telegram-bot/issues/899#issuecomment-691856766

I actually don't agree with that, many other web-frameworks are sensitive to orders like [FastAPI](https://github.com/fastapi/fastapi) and well cleared that [Order matters](https://fastapi.tiangolo.com/tutorial/path-params/#order-matters). BTW this is related on user skills and I think this is not a good strategy to keep yourselves from improvements because of maybe some users doesn't get it!
